### ISSUE TITLE
Problem in parsing with custom grammer

### ISSUE THREAD

**Issue by prakhar21** [Engelberg/instaparse#141]
I am new to Instaparse, and have started exploring few days back only.

**I have defined two functions as below:**

```
(def whitespace
  (insta/parser
    "whitespace = #'\\s+'"))

(def parser (insta/parser
                "S = word*
                <word> = NP
                <w> = #'[A-Za-z]+'
                NP = (w seperator N)+
                <N> = <'NNP'>
                <seperator> = <'_'>"
                :auto-whitespace whitespace))
```

**Input:** (parser "kohli_NNP cricket_NNP")
**Output that I want:** [:S [:NP "kohli" "cricket"]]
**Output that I am getting**: [:S [:NP "kohli"] [:NP "cricket"]]

I am not able to figure out, why is it so. Please, help me solve this.
Thanks.

**Comment by Engelberg** [comment]
This looks to me like an ambiguous grammar (I haven't actually tested it out myself, but you can likely confirm this by using the insta/oarses function to list all possible parses.  I expect that both your expected output and the output you are getting are both valid parses).

The reason the output you are getting is valid is because of that first line, `S=word*`.  Each NP is a valid word, and S permits more than one word.  So one interpretation here is that you get a bunch of words and each word is one NP.

The output you want would show up in the list of all possible parses because of the `+` in the NP definition.  So another possible interpretation is that you want one word and multiple NPs in that word.

So in this case if you remove the `*` from the first line, so it reads `S = word`, this should force the interpretation of one word with multiple NPs, and I think you'll get the proper output in this particular case.  I'm not sure what you want in general, though, so it may take deeper thought to come up with something that reflects the general behavior you want.

Calling insta/parses is the most useful tool for diagnosing ambiguity.  Calling it with the flag `:unhide :all` -- this temporarily unhides hidden things so you can better understand the structure of the parse.

**Comment by prakhar21** [comment]
@Engelberg  Thanks for the reply.
**My use case is:** Given an input ( Like, I have mentioned above ) I want to extract all the NP(Noun Phrases) from it. Right now, I am just focusing upon extracting multiple occurrence of Nouns occurring consecutively. So that, if we draw pares tree for that, I can put those consecutive occurring nouns inside a node NP.

What I am getting is, two nodes from S with label NP but I want just one node from S with label NP that has all the consecutive occurring Nouns.

Thanks

**Comment by Engelberg** [comment]
```
                "S = NP
                <w> = #'[A-Za-z]+'
                NP = (w seperator N)+
                <N> = <'NNP'>
                <seperator> = <'_'>"

Eliminating the whole `word*` piece as above sounds to me like it matches your verbal description of what you are looking for.

**Comment by Engelberg** [comment]
If there are some cases where you want it to be interpreted as multiple NPs and some where you want it to be interpreted as one NP, then you have to put some thought into how the grammar is supposed to know the difference.  If there is no way to tell the difference, then maybe ambiguity is exactly what you want, and it's great that instaparse can handle that.  Instaparse produces all viable parses and then you can add another layer of logic to try to figure out which one is the right one.

**Comment by prakhar21** [comment]
@Engelberg It's working, but I don't get it to why is it working.

My understanding is that, Given a sentence (S) which is made of zero or more words (words*), then a word can be a NP(Noun phrase), where Noun phrase is nothing but that pattern that I had given.


Where is my understanding wrong ? Please, help me out

Thanks

**Comment by Engelberg** [comment]
Yes, that is a good summary of your original version.

The particular parse that came out first was the one in which interpreted your sentence as being made up of *two* words (which is allowed by zero or more words, i.e., `words*`).  Each word was made up of a noun phrase, where noun phrase was exactly *one* instance of the pattern you had given (the `+` symbol allows one or more, so one instance is valid).

So you see, both parses are allowed by your original grammar.

**Comment by Engelberg** [comment]
If you :unhide :all, you might better be able to see why the parse matches your grammar.

**Comment by prakhar21** [comment]
@Engelberg  Thanks for helping.
But, Still I am facing issues in parsing things.

Let's say, I wish to find PP(Preposition phrases) that will be made of Preposition + NP
I have written  the below grammer:
`              "S = word
                <word> = PP | NP
                PP = w seperator P NP
                <P> = <'IN'>
                <w> = #'[A-Za-z0-9]+'
                <N> = <'NNP'> | <'NN'> | <'NNS'>
                NP = ((w seperator D)? (w seperator JJ)* w seperator N) | (w seperator N)+
                <JJ> = <'JJ'> | <'JJS'>
                <D> = <'DET'> 
                <seperator> = <'_'>"
                :auto-whitespace whitespace))
`

Problem I am facing is that, when I provide input as: (parser "in_IN the_DET great_JJ finals_NNP")
I am getting the correct output: [:S [:PP "in" [:NP "the" "great" "finals"]]]


But, when I pass: (parser "against_IN bangladesh_NNP in_IN the_DET great_JJ finals_NNP")
I am getting the error: 

```
Parse error at line 1, column 30:
against_IN bangladesh_NNP in_IN the_DET great_JJ finals_NNP
                             ^
Expected one of:
"NNS"
#"\s+"
"NN"
"NNP"
```

Basically, it should make two PP nodes with one as **against bangladesh** and another as **in the great finals**

**Comment by Engelberg** [comment]
Right now your grammar specifies that a sentence is a word which is *one* PP or NP.  To get the behavior you want you'd need to say something like `<word> = PP+ | NP`.

**Comment by prakhar21** [comment]
@Engelberg Yes, that looks good to me as well as to my parser. But, now why my parser is not able to parse sentences like **kohli_NN in_IN bengal_NNP** , it throws error saying
```
kohli_NNP in_IN bengal_NNP
             ^
Expected one of:
"NNS"
#"\s+"
"NN"
"NNP"
```
But, I have clearly define word to be either NP or PP+
It should parse by NP rule right ?

Thanks

**Comment by Engelberg** [comment]
This is doing exactly what your grammar says.
Your grammar says to expect either a series of PPs (i.e., PP+) or one NP (which can be several NNPs).

Your input string is a mixture of an NP and a PP, which is not a valid combination according to your grammar.

Again, it's hard for me to know what you really want, but the parser is doing exactly what your grammar says.  So keep thinking through the grammar and make sure it represents what you are trying to parse.

**Comment by prakhar21** [comment]
Hey @Engelberg I am trying to find out noun phrases from the given sentences.
I have written grammer as:
```
(def parser (insta/parser
                "S = NP*
                NP = ruleNP01 | ruleNP02
                ruleNP01 = (w seperator Noun)
                ruleNP02 = (w seperator D)? (w seperator J)* (w seperator Noun)
                <w> = #'[A-Za-z0-9\\-]+'
                <D> = <'DET'>
                <J> = <'JJS'> | <'JJ'>
                <Noun> = <'NN'> | <'NNP'> | <'NNS'>
                <seperator> = <'_'>"
                :auto-whitespace whitespace
                ))
```

Input: (insta/parses parser "kohli_NNP good_JJ bangladesh_NN bad_JJ india_NNP")
Output: ([:S [:NP [:ruleNP02 "kohli"]]
                   [:NP [:ruleNP02 "good" "bangladesh"]]
                   [:NP [:ruleNP02 "bad" "india"]]] 
              [:S [:NP [:ruleNP01 "kohli"]] 
              [:NP [:ruleNP02 "good" "bangladesh"]] 
              [:NP [:ruleNP02 "bad" "india"]]])


I want the 2nd **S**. But, by default it is returning me the 1st one. It's a case of ambiguity. How do I solve this ?

**Comment by Engelberg** [comment]
Try the ordered choice operator:
```
NP = ruleNP01 / ruleNP02
```

**Comment by prakhar21** [comment]
@Engelberg I have solved that problem by writing down grammer in more simpler form and avoiding regular expressions.
```

(def parser (insta/parser
                "S = NP*
                NP = ruleNP01 | ruleNP02 | ruleNP03 | ruleNP04
                ruleNP01 = (w seperator Noun)
                ruleNP02 = w seperator D w seperator Noun
                ruleNP03 = w seperator J w seperator Noun
                ruleNP04 = (w seperator D)+ (w seperator J)+ w seperator Noun
                <w> = #'[A-Za-z0-9\\-]+'
                <D> = <'DET'>
                <J> = <'JJS'> | <'JJ'>
                <Noun> = <'NN'> | <'NNP'> | <'NNS'>
                <seperator> = <'_'>"
                :auto-whitespace whitespace
                ))
```
Now, that problem is solved, all of my test cases are getting parsed correctly expect the ones where consecutive Nouns occur. There, I want them to be under same node NP by they are getting into multiple nodes of NP each having one noun.
I have tried changing `ruleNP01 = (w seperator Noun) **to** ruleNP01 = (w seperator Noun)+`
but, no effect. On the other hand if I write `ruleNP01 = (w seperator Noun w seperator Noun)
` then it matches 2 consecutive nouns. Then why is **+** not working for the same.

**Comment by Engelberg** [comment]
+ doesn't help because + also allows 1 noun, and that is consistent with your rules.
Your other form only accepts if it sees 2 nouns.

You probably need to do something to force ruleNP01 to keep matching until there are no more nouns left.  Something like:

```
ruleNP01 = (w seperator Noun)+ !(w seperator Noun)
```

**Comment by prakhar21** [comment]
@Engelberg Thank You so much. Finally every thing is working properly for at least Noun phrase extraction. Now, I will jump to Preposition phrase extraction. Please, helping me.

Thank You once again.  :+1:

**Comment by prakhar21** [comment]
Hey, @Engelberg In that NP extraction, I am trying to overlook all the words other than Nouns.
I have defined the below grammar:
```
(def parser (insta/parser
                "S = anythingElse*
                anythingElse = (w seperator UNKNOWN)+ !(w seperator Noun)
                UNKNOWN = <'VB'>
                Noun = <'NNP'> | <'NNS'>"))
```

Input: (parser "plays_VB play_VB Mike_NNP")
But I am getting error with this. Pointing to Mike_NNP

**Comment by Engelberg** [comment]
I don't think you meant to have Noun on the anythingElse line.  Probably should be UNKNOWN, depending on what you're trying to do.
